### PR TITLE
Tarred dataset support for audio classification and label models

### PR DESCRIPTION
**Pull request by fayejf**
Cleanup and continued work on previous PR #1578

- [x] Refactor classification models and label models in audio_to_label 
- [x] Add support for tarred dataset for audio_to_label
- [x] Update classification and label models 
- [x] Update VAD inference (should belong to Speaker Diarization pipelines PR but need to be used here)
- [x] Update training script, config, and instructions to use tarred dataset and improve consistency (classification and speaker model)
- [x] Update converting_to_tarred_dataset script to support label models. 
- [x] Upload test data
- [x] Add test scripts
- [x] Ok a new one. enable using the duplicate name for tarred dataloader. (webdataset doesn't allow it) 
       Will try some hacky ways and discuss with people about a cleaner / long-term way after break. 
- [x] Add instruction/warning that audio_filepath **should not contain "-sub"** cause we will need it to handle duplicate filename (different offset) samples.

Signed-off-by: fayejf <fayejf07@gmail.com>

### CODE REVIEW

File: nemo/collections/asr/data/audio_to_label.py [hunk 1]
```diff
@@ -278,3 +330,561 @@ def __getitem__(self, index):
...
+
+        self.labels = labels if labels else self.collection.uniq_labels
+        self.num_classes = len(self.labels)
+
+        self.label2id, self.id2label = {}, {}
+        for label_id, label in enumerate(self.labels):
+            self.label2id[label] = label_id
+            self.id2label[label_id] = label
+
+        for idx in range(len(self.labels[:5])):
```
Comment (titu1994) [inline nemo/collections/asr/data/audio_to_label.py:568]: Is this needed for debugging or for actual logging?

File: nemo/collections/asr/data/audio_to_label.py [hunk 2]
```diff
@@ -278,3 +330,561 @@ def __getitem__(self, index):
...
+            offset = 0
+
+        # Convert audio bytes to IO stream for processing (for SoundFile to read)
+        audio_filestream = io.BytesIO(audio_bytes)
+        features = self.featurizer.process(
+            audio_filestream,
+            offset=offset,
+            duration=manifest_entry.duration,
+            trim=self.trim,
+            # orig_sr=manifest_entry.orig_sr,
```
Comment (titu1994) [inline nemo/collections/asr/data/audio_to_label.py:677]: Why is this commented out?

File: nemo/collections/asr/data/audio_to_label_dataset.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    Returns:
+        An instance of AudioToClassificationLabelDataset.
+    """
+    dataset = audio_to_label.AudioToClassificationLabelDataset(
+        manifest_filepath=config['manifest_filepath'],
+        labels=config['labels'],
+        featurizer=featurizer,
+        max_duration=config.get('max_duration', None),
+        min_duration=config.get('min_duration', None),
+        trim=config.get('trim_silence', True),
```
Comment (titu1994) [inline nemo/collections/asr/data/audio_to_label_dataset.py:42]: Default trim silence should be False

File: nemo/collections/asr/data/audio_to_label_dataset.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    Returns:
+        An instance of AudioToSpeechLabelDataset.
+    """
+    dataset = audio_to_label.AudioToSpeechLabelDataset(
+        manifest_filepath=config['manifest_filepath'],
+        labels=config['labels'],
+        featurizer=featurizer,
+        max_duration=config.get('max_duration', None),
+        min_duration=config.get('min_duration', None),
+        trim=config.get('trim_silence', True),
```
Comment (titu1994) [inline nemo/collections/asr/data/audio_to_label_dataset.py:67]: Default trim silence should be False

File: nemo/collections/asr/data/audio_to_label_dataset.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    """
+    dataset = audio_to_label.TarredAudioToClassificationLabelDataset(
+        audio_tar_filepaths=config['tarred_audio_filepaths'],
+        manifest_filepath=config['manifest_filepath'],
+        labels=config['labels'],
+        featurizer=featurizer,
+        shuffle_n=shuffle_n,
+        max_duration=config.get('max_duration', None),
+        min_duration=config.get('min_duration', None),
+        trim=config.get('trim_silence', True),
```
Comment (titu1994) [inline nemo/collections/asr/data/audio_to_label_dataset.py:106]: Default trim silence should be False

File: nemo/collections/asr/data/audio_to_label_dataset.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    """
+    dataset = audio_to_label.TarredAudioToSpeechLabelDataset(
+        audio_tar_filepaths=config['tarred_audio_filepaths'],
+        manifest_filepath=config['manifest_filepath'],
+        labels=config['labels'],
+        featurizer=featurizer,
+        shuffle_n=shuffle_n,
+        max_duration=config.get('max_duration', None),
+        min_duration=config.get('min_duration', None),
+        trim=config.get('trim_silence', True),
```
Comment (titu1994) [inline nemo/collections/asr/data/audio_to_label_dataset.py:144]: Default trim silence should be False

File: nemo/collections/asr/models/classification_models.py [hunk 7]
```diff
@@ -138,9 +151,36 @@ def transcribe(self, paths2audio_files: List[str], batch_size: int = 4, logprobs
             logging.set_verbosity(logging_level)
         return labels
 
+    def prepare_manifest(self, config):
```
Comment (titu1994) [inline nemo/collections/asr/models/classification_models.py:154]: Add the pytorch lightning decorator to run this **only on device=0**.

Also add documentation. Is this meant to be run prior to train? Or in general? If its general, remove it from Model and put it in a helper script.

File: nemo/collections/asr/models/classification_models.py [hunk 8]
```diff
@@ -138,9 +151,36 @@ def transcribe(self, paths2audio_files: List[str], batch_size: int = 4, logprobs
...
+            'time_length': config['time_length'],
+        }
+        results = p.starmap(write_manifest_data, zip(input_audios, repeat(args_func)))
+        p.close()
+
+        if os.path.exists(manifest_vad_input):
+            logging.info("The prepared manifest file exists. Overwriting!")
+            os.remove(manifest_vad_input)
+
+        with open(manifest_vad_input, 'a') as fout:
```
Comment (titu1994) [inline nemo/collections/asr/models/classification_models.py:174]: Is there a necessity to write under DDP? Cant we have a script that preprocesses this outside of model?

File: nemo/collections/asr/parts/vad_utils.py [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+
+    except Exception as e:
+        err_file = "error.log"
+        with open(err_file, 'w') as fout:
+            fout.write(file + ":" + str(e))
+
+    return res
+
+
+def get_status(data):
```
Comment (titu1994) [inline nemo/collections/asr/parts/vad_utils.py:95]: too generic function name - how about `get_vad_stream_status`

File: nemo/collections/asr/parts/vad_utils.py [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+import json
+import logging
+import os
+
+import librosa
+import numpy as np
+import pandas as pd
+
+
+def write_manifest_data(file, args_func):
```
Comment (titu1994) [inline nemo/collections/asr/parts/vad_utils.py:24]: too generic function name - how about `write_vad_manifest_data`

File: nemo/collections/asr/parts/vad_utils.py [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+            elif data[i] == data[i - 1] and data[i] == data[i + 1]:
+                status[i] = 'next'
+            elif data[i] == data[i - 1] and data[i] != data[i + 1]:
+                status[i] = 'end'
+            else:
+                status[i] = 'single'
+    return status
+
+
+def gen_overlap_seq(frame_filepath, per_args):
```
Comment (titu1994) [inline nemo/collections/asr/parts/vad_utils.py:121]: too generic function name - how about `generate_overlap_vad_seq`

File: nemo/collections/asr/parts/vad_utils.py [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+
+        round_final = np.round(preds, 4)
+        np.savetxt(overlap_filepath, round_final, delimiter='\n')
+        logging.info(f"Finished! {overlap_filepath}!")
+
+    except Exception as e:
+        raise (e)
+
+
+def gen_seg_table(frame_filepath, per_args):
```
Comment (titu1994) [inline nemo/collections/asr/parts/vad_utils.py:206]: too generic function name - and dont use short forms like gen and seg

File: nemo/collections/asr/parts/vad_utils.py [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+    state_list.append(current_sate)
+
+    seg_table = pd.DataFrame({'start': start_list, 'end': end_list, 'vad': state_list})
+
+    save_name = name + ".txt"
+    save_path = os.path.join(out_dir, save_name)
+    seg_table.to_csv(save_path, sep='\t', index=False, header=False)
+
+
+def write_manifest(vad_directory, audio_directory, manifest_file):
```
Comment (titu1994) [inline nemo/collections/asr/parts/vad_utils.py:255]: too generic function name

File: nemo/collections/asr/parts/vad_utils.py [hunk 14]
```diff
@@ -0,0 +1,272 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
Comment (titu1994) [inline nemo/collections/asr/parts/vad_utils.py:1]: The functions in this script are too generic. If you would like to keep the names, please import vad_utils such that all methods are called as `vad_utils.XYZ()`.

File: scripts/convert_to_tarred_audio_dataset.py [hunk 15]
```diff
@@ -68,21 +72,44 @@ def create_shard(entries, target_dir, new_entries, shard_id):
...
-        new_entry = {
-            'audio_filepath': squashed_filename,
-            'duration': entry['duration'],
-            'text': entry['text'],
-            'shard_id': shard_id,  # Keep shard ID for recordkeeping
-        }
+        if squashed_filename not in count:
+            tar.add(entry['audio_filepath'], arcname=squashed_filename)
+
+        if entry['label']:
```
Comment (titu1994) [inline scripts/convert_to_tarred_audio_dataset.py:86]: What if `label`  key doesnt exist in entry - ASR manifests dont have label but text instead? First check for key, then access it.

**Review by titu1994 — CHANGES_REQUESTED** [review]
Overall it looks great.

There is a major concern with why we are performing data writing inside of a ModelPT class - please avoid it at all costs unless there is no other choice at all. 

Some other minor things are function names and documentation.

File: nemo/collections/asr/data/audio_to_label.py (hunk 2 above)
Comment (fayejf) [reply nemo/collections/asr/data/audio_to_label.py]: manifest_entry  doesn't have orgin_sr (for now) in classification and label models

File: scripts/convert_to_tarred_audio_dataset.py (hunk 15 above)
Comment (fayejf) [reply scripts/convert_to_tarred_audio_dataset.py]: good catch!!!

**Review by titu1994 — APPROVED** [review]
LGTM, minor comments here and there. @jbalam-nv for final review before merge.

File: nemo/collections/asr/models/classification_models.py [hunk 16]
```diff
@@ -43,6 +44,15 @@ class EncDecClassificationModel(ASRModel, Exportable):
     """Encoder decoder CTC-based models."""
```
Comment (jbalam-nv) [inline nemo/collections/asr/models/classification_models.py:44]: This description needs to be modified

File: nemo/collections/asr/models/classification_models.py [hunk 17]
```diff
@@ -150,58 +158,102 @@ def _setup_dataloader_from_config(self, config: Optional[Dict]):
...
-                featurizer=featurizer,
-                max_duration=config.get('max_duration', None),
-                min_duration=config.get('min_duration', None),
-                trim=config.get('trim_silence', True),
-                load_audio=config.get('load_audio', True),
+                config=config,
+                shuffle_n=shuffle_n,
+                global_rank=self.global_rank,
+                world_size=self.world_size,
+                augmentor=augmentor,
```
Comment (jbalam-nv) [inline nemo/collections/asr/models/classification_models.py:185]: augmentor is passed as an argument here but we do not do anything with it in get_tarred_classification_label_dataset()

**Review by jbalam-nv — CHANGES_REQUESTED** [review]
Support for augmentor is the big change requested, otherwise minor changes to comments and grammar.

File: nemo/collections/asr/data/audio_to_label.py [hunk 18]
```diff
@@ -11,51 +11,279 @@
...
+        count (Dict): Dict of wav files {'A' : 2, ..., 'S':10}
+    """
+    count = dict()
+    for i in manifest_file_id:
+        audio_filename = i.split("-sub")[0]
+        count[audio_filename] = count.get(audio_filename, 0) + 1
+    return count
+
+
+def _speech_collate_fn(batch, pad_id):
```
[the remark this answers is not in the document]
Comment (fayejf) [reply nemo/collections/asr/data/audio_to_label.py:78]: Discussed in slack. Maybe it’s better to keep where they are used for now. And quote from Som “Its possible for cyclic dependency to come in the future by centralizing it”

File: nemo/collections/asr/data/audio_to_label.py [hunk 19]
```diff
@@ -278,3 +355,550 @@ def __getitem__(self, index):
...
+        normalize_audio (bool): Whether to normalize audio signal. 
+            Defaults to False.
+    """
+
+    def __init__(
+        self,
+        *,
+        manifest_filepath: str,
+        labels: List[str],
+        featurizer,
```
[the remark this answers is not in the document]
Comment (fayejf) [reply nemo/collections/asr/data/audio_to_label.py:436]: see below comment plz

**Comment by fayejf** [comment]
Augmentor is passed to featurizer in classification and label model.  And featurizer is passed to `AudioLabelDataset`/`AudioToSpeechLabelDataSet` by `get_classification_label_dataset/get_speech_label_dataset`.  Well `audio_to_text` process augmentor and featurizer inside `_AudioTextDataset` 

I think expose featurzier (pass in augmentor)  (as in classification models ) makes more sense since one day we may end up using different featurizer. And its clearer what exactly is going on for augmentation

well I won’t change `audio_to_text` here cause it doesn’t belong to it. but leave here a note and might change it later.

File: examples/asr/vad_infer.py [hunk 20]
```diff
@@ -63,6 +65,16 @@ def main():
...
     parser.add_argument("--time_length", type=float, default=0.63)
     parser.add_argument("--shift_length", type=float, default=0.01)
+    parser.add_argument("--normalize_audio", type=bool, default=False)
+    parser.add_argument("--num_workers", type=float, default=20)
+    parser.add_argument("--split_duration", type=float, default=400)
+    parser.add_argument(
+        "--dont_auto_split",
+        default=False,
+        action='store_true',
+        help="Whether to automatically split manifest entry by split_duration to avoid potential issue.",
```
Comment (redoctopus) [inline examples/asr/vad_infer.py:75]: Add more info on the "potential issue"?

Comment (redoctopus) [inline examples/asr/vad_infer.py:75]: Add more info on the "potential issue"?

File: nemo/collections/asr/data/audio_to_label.py [hunk 21]
```diff
@@ -11,51 +11,271 @@
...
+    """normalize signal
+    Args: 
+        signal(FloatTensor): signal to be normalized.
+    """
+    signal_minusmean = signal - signal.mean()
+    return signal_minusmean / signal_minusmean.abs().max()
+
+
+def count_occurence(manifest_file_id):
+    """Cound number of wav files in Dict manifest_file_id. Use for _TarredAudioToLabelDataset.
```
Comment (redoctopus) [inline nemo/collections/asr/data/audio_to_label.py:57]: Minor typo: s/Cound/Count

Comment (fayejf) [reply nemo/collections/asr/data/audio_to_label.py:57]: haha good catch 😉

**Review by jbalam-nv — APPROVED** [review]
LGTM
